Catch invalid aliases at merging rather than subtyping

clementblaudeau opened the pull request:
This PR proposes a refactor of the handling of invalid aliases introduced by module constraints

# Context

Module aliases to non-alisable paths (functor applications, functor parameters, recursive modules inside the recursive knot) can be introduced via destructive module constraints `... with module X := P`, as described in #6376. For instance: 
```ocaml
module X = struct end 
module F (_:sig end) = struct end
module type T = (sig 
  module X0 : sig end 
  module X1 = X0 
end) with module X0 := F(X)
```
Could incorrectly result in `module type T = sig module X1 = F(X) end`. This was addressed by 82e2f35, which adds a check in *subtyping*, when comparing alias signatures, that neither is ill-formed. The code above result in the following subtyping error:

```ocaml
Error: Module type declarations do not match:
         module type T = sig module X1 = F(X) end
       does not match
         module type T = sig module X1 = F(X) end
       At position module type T = <here>
       Module types do not match:
         sig module X1 = F(X) end
       is not equal to
         sig module X1 = F(X) end
       At position module type T = sig module X1 : <here> end
       Module F(X) cannot be aliased
``` 

There are three issues with the current situation: 
1. the check is not done at the right moment, it should prevent the introduction of invalid aliases rather than their subtyping with another alias
   <details>
     <summary>Details</summary>
     
   The error is caught when the invalid signature reaches the top-level, as there is a subtyping check of the signature against itself. In the "mean time" between the introduction of the signature and the toplevel, the invalid signature is in the typing environment, which is an open door to issues. Typically, the following code succeeds: 
   ```ocaml 
   module X = struct end
   module F (_:sig end) = struct type t end
   module ShoudFail : sig end = struct
     module type T = (sig
       module X0 : sig end
       module X1 = X0
     end)
       with module X0 := F(X)
   end
   ``` 
   Indeed, the invalid signature is accepted and then removed by an ascription, so the subtyping check on the content of `T` never triggers. Similarly, consider the following code:
   
   ```ocaml 
   module X = struct end
   module F (_:sig end) = struct type t end
   module M = struct
     module type Should_fail_here = (sig
       module X0 : sig end
       module X1 = X0
     end)
       with module X0 := F(X)
   
     (* here, the env contains an invalid signature *)
     module Should_not_fail_here : sig type t = int end =
     struct type t = bool end
   end 
   ```
   It fails at `Should_not_fail_here` rather than at `Should_fail_here`. I have not looked to hard for more serious errors resulting in having an invalid alias in the typing env, but it is likely usable to (at least) crash to typechecker.
   </details>
2. the subtyping error is misleading
3. not all cases of invalid aliases are caught
   <details>
     <summary>Details</summary>
     
   * Aliases to functor parameters are accepted when they should not:
   ```ocaml
   module F (Y:sig end) = struct
     module type T = sig
       module X0 : sig end
       module X1 = X0
     end with module X0 := Y
   end
   (* module F : (Y : sig end) -> sig module type T = sig module X1 = Y end end *)
   ```
   * Aliases to recursive modules are accepted inside the recursive knot:
   ```ocaml
   module rec Xrec : sig end = struct
     module type T = sig
       module X0 : sig end
       module X1 = X0
     end with module X0 := Xrec
   end
   (* module rec Xrec : sig end *)
   ```
   
   As for 1, it may probably be used to trigger more serious bugs. 
   </details>

# This PR

This PR addresses those issues as follows: 
* It introduces a new check when doing destructive module constraints with a non-aliasable path `check_invalid_aliases`, along with a new error case `With_creates_invalid_aliases`
* it removes the check done in the subtyping for validity of aliases
* It lightly refactors `check_usage_of_path_of_substituted_item` and `path_is_prefix` to accomodate for the extra check (regardless of whether the substitution is deep or not)
* it adds tests (with shallow/deep substitutions, with functor parameters, applications, recursive modules, etc.)

This PR should not affect performance, as it re-uses the visit of the signature done for the other wellformedness checks (`check_usage_of_path_of_substituted_item`). It is almost backward compatible, as it will only prevent aliases that were invalid and not usable anyway (i.e. the incompatibility was out of spec).

However, if there is another way to introduce invalid aliases that I'm not aware of, they will not be caught by subtyping. I believe that it is ok, as the role of subtyping is not to safeguard invalid inference. 

# Examples 

Now, the above example fails with the following error message: 
```ocaml 
module X = struct end 
module F (_:sig end) = struct end
module type T = (sig 
  module X0 : sig end 
  module X1 = X0 
end) with module X0 := F(X)
```
```
Error: In this with constraint, replacing X0 by F(X) would 
       introduce an invalid alias (at X1) .
```

For other examples, see the [added test cases](https://github.com/ocaml/ocaml/pull/14066/files#diff-16475d6e2475e142459277f67dd1087fc001f8d9cf384c35c57c3eb49d67b6dc)

clementblaudeau committed:
Add tests for introduction of invalid aliases via module constraints

clementblaudeau committed:
Catch invalid aliases at module constraints, not at subtyping

clementblaudeau committed:
Remove the unreachable "Invalid alias" subtyping error

clementblaudeau committed:
changelog

Octachron commented on typing/typemod.ml:
```diff
@@ -344,28 +346,60 @@ let check_usage_of_path_of_substituted_item paths ~loc ~lid env super =
+          | Some invalid_alias_path ->
+              check_invalid_aliases paths ~loc env iterator invalid_alias_path
+        in
+        iterator.Btype.it_signature iterator' sg
```
I am wondering if we want to move the testing of conditions to `check_usage_of_path_of_substituted_item` and `check_invalid_aliases` in order to reduce the combination of the two iterator transformer to  just
```ocaml
let iterator = 
       base_iterator 
  |> check_usage_of_path_of_substituted_item paths ~loc ~lid env
  |> check_invalid_aliases paths ~loc env iterator invalid_alias_path
in
iterator.Btype.it_signature iterator' sg
```

Octachron commented on typing/typemod.ml:
```diff
@@ -3440,6 +3478,15 @@ let report_error ~loc _env = function
+      Location.errorf ~loc
+        "@[<v>\
+         @[In this %a constraint, replacing %a by %a would @ \
+         introduce an invalid alias (at %a)@].@]"
```
Opening two formatting boxes consecutively with no contents in-between is equivalent to opening only the last box. And in fact no formatting box is needed here. Moreover, it is generally a good idea to have a break hint before arbitrarily large contents. Finally the name of the problematic alias is probably too important to let between parentheses:

```ocaml
Location.errorf ~loc
  "In this %a constraint,@ replacing %a@ by %a@ would@ \
    introduce an invalid alias@ at %a."
```

Octachron commented on typing/typemod.ml:
```diff
@@ -249,21 +250,24 @@ let rec iter_path_apply p ~f =
+    | `Ok (ident1, l1), `Ok (ident2, l2)
+      when Ident.same ident1 ident2 ->
+        list_is_prefix l1 ~prefix:l2
+    | _,_ -> false
```
This is only a cosmetic change, isn't it?

Octachron commented on typing/typemod.ml:
```diff
@@ -249,21 +250,24 @@ let rec iter_path_apply p ~f =
-       String.equal s1 s2 && list_is_strict_prefix t1 ~prefix:t2
+    | s1 :: t1, s2 :: t2 when (String.equal s1 s2) ->
+        list_is_prefix t1 ~prefix:t2
+    | _, _ -> false
```
Same as below, this is a cosmetic change (except for the first line).

Octachron commented:
I agree that avoiding the creation of invalid aliases is a more robust fix than failing later on the interface inclusion check

clementblaudeau committed:
Fix error message (remove useless boxes + add break hints)

clementblaudeau committed:
Revert cosmetic changes

clementblaudeau committed:
Refactor the computation of iterators

clementblaudeau commented:
Thank you for the review. I've addressed your comments in the last three commits. Once the review is finished, I can rebase the PR to have only a commit that adds the (failing) tests and a commit that does the fixing.

Octachron commented on typing/typemod.ml:
```diff
@@ -344,28 +360,53 @@ let check_usage_of_path_of_substituted_item paths ~loc ~lid env super =
+    introduction of invalid aliases *)
+let check_usage_after_substitution env ~loc ~lid paths ?(invalid_alias=None) sg=
+  match paths, invalid_alias with
+  (* Shallow substitution without introduction of an invalid alias *)
```
The comment could be extended to ` Shallow substitution without introduction of an invalid alias : nothing to check` to make it explicit that this branch is here to avoid doing a no-op traversal.

Octachron reviewed (approved):
I agree that avoiding the creation of invalid module types is an improvement, and I find overall the new code simpler; and ready to be merged. 

Thanks for the fix !

clementblaudeau committed:
Update comment
